まず最初にpythonのみで

In [1]:
import math

In [2]:
import numpy as np

In [3]:
S0=36
T=1.0
r=0.06
sigma=0.2

In [4]:
def simulate_tree(M):
    dt=T/M
    u=math.exp(sigma*math.sqrt(dt))
    d=1/u
    S=np.zeros((M+1,M+1))
    S[0,0]=S0
    z=1
    for t in range(1,M+1):
        for i in range(z):
            S[i,t]=S[i,t-1]*u
            S[i+1,t]=S[i,t-1]*d
        z+=1
    return S

In [5]:
np.set_printoptions(formatter={'float':
                               lambda x: '%6.2f' %x })
#formatter, prints the matrix in orderly fashion, 
#in this case each element is printed in float on 6.2f s(up to two decimals,for 6 spaces)

In [6]:
simulate_tree(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [  0.00,  32.57,  36.00,  39.79,  43.97],
       [  0.00,   0.00,  29.47,  32.57,  36.00],
       [  0.00,   0.00,   0.00,  26.67,  29.47],
       [  0.00,   0.00,   0.00,   0.00,  24.13]])

In [7]:
%time simulate_tree(500)

CPU times: user 138 ms, sys: 5.85 ms, total: 144 ms
Wall time: 156 ms


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [  0.00,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [  0.00,   0.00,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.00,   0.41]])

次はnumpyをしっかり使ってみる

In [8]:
M=4

In [9]:
up=np.arange(M+1)
up=np.resize(up,(M+1,M+1))
up

array([[0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4]])

In [10]:
down=up.T*2
down

array([[0, 0, 0, 0, 0],
       [2, 2, 2, 2, 2],
       [4, 4, 4, 4, 4],
       [6, 6, 6, 6, 6],
       [8, 8, 8, 8, 8]])

In [11]:
up-down

array([[ 0,  1,  2,  3,  4],
       [-2, -1,  0,  1,  2],
       [-4, -3, -2, -1,  0],
       [-6, -5, -4, -3, -2],
       [-8, -7, -6, -5, -4]])

In [12]:
dt=T/M

In [13]:
S0*np.exp(sigma*math.sqrt(dt)*(up-down))

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [ 29.47,  32.57,  36.00,  39.79,  43.97],
       [ 24.13,  26.67,  29.47,  32.57,  36.00],
       [ 19.76,  21.84,  24.13,  26.67,  29.47],
       [ 16.18,  17.88,  19.76,  21.84,  24.13]])

In [14]:
def simulate_tree_np(M):
    dt=T/M
    up=np.arange(M+1)
    up=np.resize(up,(M+1,M+1))
    down=up.transpose()*2
    S =S0*np.exp(sigma*math.sqrt(dt)*(up - down))
    return S

In [15]:
simulate_tree_np(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [ 29.47,  32.57,  36.00,  39.79,  43.97],
       [ 24.13,  26.67,  29.47,  32.57,  36.00],
       [ 19.76,  21.84,  24.13,  26.67,  29.47],
       [ 16.18,  17.88,  19.76,  21.84,  24.13]])

In [16]:
%time simulate_tree_np(500)

CPU times: user 9.33 ms, sys: 5.31 ms, total: 14.6 ms
Wall time: 13.7 ms


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [ 35.36,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [ 34.73,  35.05,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.41,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.40,   0.41,   0.41]])

cf)pythonのみの場合
CPU times: user 116 ms, sys: 2.95 ms, total: 118 ms
Wall time: 118 ms

次はNumbaで

In [17]:
import numba

In [18]:
simulate_tree_nb=numba.jit(simulate_tree)

In [19]:
simulate_tree_nb(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [  0.00,  32.57,  36.00,  39.79,  43.97],
       [  0.00,   0.00,  29.47,  32.57,  36.00],
       [  0.00,   0.00,   0.00,  26.67,  29.47],
       [  0.00,   0.00,   0.00,   0.00,  24.13]])

In [20]:
%time simulate_tree_nb(500)

CPU times: user 482 µs, sys: 24 µs, total: 506 µs
Wall time: 544 µs


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [  0.00,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [  0.00,   0.00,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.00,   0.41]])

In [21]:
%timeit simulate_tree_nb(500)

295 µs ± 29.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


%time simulate_tree(500)
CPU times: user 116 ms, sys: 2.95 ms, total: 118 ms
Wall time: 118 ms

%time simulate_tree_np(500)
CPU times: user 15.6 ms, sys: 9.59 ms, total: 25.2 ms
Wall time: 31.3 ms

%time simulate_tree_nb(500)
CPU times: user 498 µs, sys: 7 µs, total: 505 µs
Wall time: 735 µs
 →すごい！

cythonを考えてみよう

In [22]:
%load_ext Cython

In [23]:
%reload_ext Cython

In [57]:
%%cython 
import numpy as np
cimport cython
from libc.math cimport exp, sqrt
cdef float S0=36.
cdef float T=1.0
cdef float r=0.06
cdef float sigma=0.2
def simulate_tree_cy(int M):
    cdef int z, t, i
    cdef float dt, u, d,
    cdef float[:,:] S=np.zeros((M+1,M+1),
                               dtype=np.float32)
    dt=T/M
    u=exp(sigma*sqrt(dt))
    d=1/u
    S[0,0]=S0
    z=1
    for t in range(1,M+1):
        for i in range(z):
            S[i,t]=S[i,t-1]*u
            S[i+1,t]=S[i,t-1]*d
        z+=1
    return np.array(S)

In [58]:
simulate_tree_cy(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [  0.00,  32.57,  36.00,  39.79,  43.97],
       [  0.00,   0.00,  29.47,  32.57,  36.00],
       [  0.00,   0.00,   0.00,  26.67,  29.47],
       [  0.00,   0.00,   0.00,   0.00,  24.13]], dtype=float32)

In [59]:
%time simulate_tree_cy(500)

CPU times: user 484 µs, sys: 88 µs, total: 572 µs
Wall time: 409 µs


array([[ 36.00,  36.32,  36.65, ..., 3095.77, 3123.59, 3151.65],
       [  0.00,  35.68,  36.00, ..., 3040.89, 3068.21, 3095.77],
       [  0.00,   0.00,  35.36, ..., 2986.97, 3013.81, 3040.89],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.00,   0.41]],
      dtype=float32)

In [60]:
%timeit S=simulate_tree_cy(500)

234 µs ± 21.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


monte carlo simulation

python版から

In [61]:
M=100#（time horizon）=1を100に分割(discrete setup)
I=50000 #試行回数を50000回に

In [62]:
def mcs_simulation_py(p):
    (M, I)=p#pは行列の形で入力する必要がある
    dt=T/M
    S=np.zeros((M+1,I))
    S[0]=S0
    rn=np.random.standard_normal(S.shape)  
    for t in range(1,M+1):  
        for i in range(I):  
            S[t,i]= S[t-1,i] * math.exp((r-sigma ** 2/2)*dt+
                                         sigma*math.sqrt(dt)*rn[t,i])
    return S

In [63]:
mcs_simulation_py((4,5)) 

array([[ 36.00,  36.00,  36.00,  36.00,  36.00],
       [ 37.58,  35.17,  39.31,  35.72,  36.91],
       [ 40.52,  27.15,  43.17,  34.03,  39.21],
       [ 42.08,  31.72,  36.59,  37.01,  39.10],
       [ 35.39,  31.92,  39.94,  33.38,  45.09]])

In [64]:
%time S=mcs_simulation_py((M, I)) 

CPU times: user 6.62 s, sys: 22.4 ms, total: 6.64 s
Wall time: 6.67 s


In [65]:
%time A=mcs_simulation_py((M, I)) 

CPU times: user 8.26 s, sys: 81 ms, total: 8.35 s
Wall time: 8.99 s


In [66]:
S[-1].mean()#terminal t=1における各パスでの株価の平均をとる

38.24430869001903

In [71]:
S0*math.exp(r*T)#black-scholesに従う場合のterminalでの価格　理論値

38.22611567563295

In [72]:
K=40.#strike priceを40に設定して

In [73]:
C0=math.exp(-r*T)*np.maximum(K-S[-1],0).mean()#評価式

In [74]:
C0

3.821123874266269

次はnumpy仕様

In [75]:
def mcs_simulation_np(p):
    (M,I)=p
    dt=T/M
    S=np.zeros((M+1,I))
    S[0]=S0
    rn=np.random.standard_normal(S.shape)
    for t in range(1,M+1):
        S[t]=S[t-1]*np.exp((r-sigma **2/2)*dt+
                                         sigma*math.sqrt(dt)*rn[t])#さっきとは違い一度に行ごと扱うことができる
    return S

In [76]:
mcs_simulation_np((5,6))

array([[ 36.00,  36.00,  36.00,  36.00,  36.00,  36.00],
       [ 38.62,  33.03,  35.27,  35.21,  40.02,  34.54],
       [ 39.79,  33.93,  34.80,  40.30,  38.36,  33.44],
       [ 47.37,  34.04,  31.72,  45.56,  41.51,  34.62],
       [ 42.21,  39.03,  35.81,  39.81,  42.86,  37.66],
       [ 43.23,  40.31,  41.07,  45.54,  37.77,  37.10]])

In [81]:
%time S=mcs_simulation_np((M,I))

CPU times: user 208 ms, sys: 8.15 ms, total: 216 ms
Wall time: 187 ms


%time S = mcs_simulation_py((M, I))
CPU times: user 5.67 s, sys: 57.4 ms, total: 5.72 s
Wall time: 5.77 s
→すごい

In [82]:
S[-1].mean()

38.210591295364665

In [83]:
%timeit S=mcs_simulation_np((M,I))

199 ms ± 13.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


Numbaを使った場合

In [84]:
mcs_simulation_nb=numba.jit(mcs_simulation_py)

In [85]:
%time S=mcs_simulation_nb((M,I))

CPU times: user 524 ms, sys: 13.9 ms, total: 538 ms
Wall time: 411 ms


In [86]:
%time S=mcs_simulation_nb((M,I))

CPU times: user 196 ms, sys: 6.74 ms, total: 203 ms
Wall time: 204 ms


In [87]:
S[-1].mean()

38.24100315363438

In [88]:
C0=math.exp(-r*T)*np.maximum(K-S[-1],0).mean()

In [89]:
C0

3.82484287214865

In [90]:
%timeit S= mcs_simulation_nb((M,I))

191 ms ± 7.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


%time S = mcs_simulation_py((M, I))
CPU times: user 5.67 s, sys: 57.4 ms, total: 5.72 s 
Wall time: 5.77 s 

%time S=mcs_simulation_np((M,I))
CPU times: user 219 ms, sys: 29.8 ms, total: 249 ms
Wall time: 224 ms
→もっとすごい

今度はCythonで

In [104]:
%%cython
import numpy as np
cimport numpy as np
cimport cython
from libc.math cimport exp, sqrt
cdef float S0 = 36.
cdef float T = 1.0
cdef float r = 0.06
cdef float sigma = 0.2
@cython.boundscheck(False)#気楽に計算してもらうようにcythonに命令する（高速化のためか）
@cython.wraparound(False)#ベクトルを-表記で逆から参照する機能を無効化（高速化のためか、つかわないし）
def mcs_simulation_cy(p):
    cdef int M, I
    M, I = p
    cdef int t, i
    cdef float dt = T / M
    cdef double[:, :] S = np.zeros((M + 1, I))
    cdef double[:, :] rn = np.random.standard_normal((M + 1, I))
    S[0] = S0
    for t in range(1, M + 1):
        for i in range(I):
            S[t, i] = S[t-1, i] * exp((r - sigma ** 2 / 2) * dt +
                                                   sigma * sqrt(dt) * rn[t, i])
    return np.array(S)

In [105]:
%time s=mcs_simulation_cy((M,I))

CPU times: user 224 ms, sys: 43.9 ms, total: 268 ms
Wall time: 249 ms


In [107]:
s[-1].mean()

38.20938953558063

In [94]:
%timeit S=mcs_simulation_cy((M,I))

210 ms ± 15.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


最後はmultiprossesing

In [112]:
import multiprocessing as mp

In [113]:
pool = mp.Pool(processes=4)

In [120]:
p=20

In [141]:
np.hstack(pool.map(mcs_simulation_np, 2*[(3, int(4/2))]))

array([[ 36.00,  36.00,  36.00,  36.00],
       [ 39.96,  37.81,  43.99,  36.63],
       [ 50.70,  30.21,  55.85,  46.33],
       [ 61.42,  25.26,  63.23,  40.30]])

In [119]:
%timeit S=np.hstack(pool.map(mcs_simulation_np, p*[(M, int(I/p))]))
#map関数,mcs_simulation_npにp*[(M, int(I/p))]を突っ込む　　のをpoolして　hstack
#Iは試行回数、(I/p)列分をp回行うことを4processで分割する
#ex)10000回やりたいときは （10000/20)=50の列を20回行うことを4分割して行う

223 ms ± 11.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [116]:
%timeit S=np.hstack(pool.map(mcs_simulation_nb, p*[(M, int(I/p))]))

237 ms ± 22.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [117]:
%timeit S=np.hstack(pool.map(mcs_simulation_cy, p*[(M, int(I/p))]))

225 ms ± 11.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


おわり